src/main/kotlin/org/rust/lang/core/type/TypeInference.kt

```diff
@@ -0,0 +1,35 @@
+package org.rust.lang.core.type
+
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.parentOfType
+import org.rust.lang.utils.psiCached
+
+val RustExpr.inferredType: RustResolvedType by psiCached {
+    when (this) {
+        is RustPathExpr -> {
+            val target = path.reference.resolve()
+            when (target) {
+                is RustSelfArgument -> {
+                    val impl = target.parentOfType<RustImplItem>()
+                    impl?.type?.resolvedType ?: RustUnknownType(manager)
+                }
+                else -> RustUnknownType(manager)
+            }
+        }
+        else -> RustUnknownType(manager)
```

Comment:
I'd elaborate on that rule: we're aligning only those `->` being placed on the adjacent lines.